### ISSUE TITLE
CRAN bound?

### ISSUE THREAD

**Issue by sckott** [wch/harbor#5]
@wch any plans to send this to CRAN?  I was thinking of getting `analogsea` to CRAN soon. We don't use `harbor` yet, but we could more easily if on CRAN

**Comment by wch** [comment]
No plans to anytime soon... If you want to clean it up, please feel free to
do so. :)

**Comment by sckott** [comment]
ok, I'm just getting back into `analogsea` work, so I'll go over the `harbor` related stuff soon

**Comment by MarkEdmondson1234** [comment]
If it just needs updates to the documentation so it passes CRAN checks, I can help with that.

**Comment by hrbrmstr** [comment]
Piling on here to see how I might help get this into CRAN and to suggest that it doesn't need to support `boot2docker` anymore with Docker being a first class citizen on macOS & Windows now. I freely admit this is selfish since I'm using it in https://github.com/hrbrmstr/splashr to install, start & stop a Splash server, but in doing so realized that this could be an excellent model for other middleware engines (like Splash, Selenium, phantomjs, etc) to keep those things separate from the local operating system and abstract the complexity away from R users.

So, I'm here to help and willing to mod/clean, but don't want to stop over any other work in the other two forks, so just lemme know what I can do to help get this puppy in CRAN.

**Comment by MarkEdmondson1234** [comment]
I did the copy-and-credit-at-will approach for my CRAN submission, maybe added an extra function or two but would prefer pulling that out into this package on CRAN

**Comment by wch** [comment]
I don't plan on spending time on harbor in the foreseeable future, so if you guys want to become admins on the project and do the release/maintenance, let me know.

**Comment by hrbrmstr** [comment]
yes pls :-)

**Comment by wch** [comment]
@hrbrmstr Done. Before you release, please feel free to make yourself the maintainer (AKA `cre`) in the DESCRIPTION file.

**Comment by hrbrmstr** [comment]
Cool. @sckott @MarkEdmondson1234 I'm prbly going to be pushing "breaking" changes soon (next 24-48 hours). I'm switching to `sys` and adding some more general Docker mgmt commands. It may not impact anything y'all are doing but wanted to make sure you both knew that. There are a handful of folks watching and Rich has a somewhat active fork, so hopefully anyone else "depending" on this will see what's going on.

**Comment by sckott** [comment]
sounds good - was going to move to using this in `analogsea` - but never got around to this, so happy someone is doing it

**Comment by cboettig** [comment]
Any updates to sending this onto CRAN?  It would work nicely I think with my little wrapper to docker-machine (https://github.com/cboettig/dockermachine) to allow R users to script deploys of long-running jobs to the cloud without having to leave the R interface

**Comment by sckott** [comment]
dont know, thoughts @hrbrmstr ?

**Comment by hrbrmstr** [comment]
I'd need to give it a few more test on Travis, but Appveyor can't do
docker-ish test (in any good/real way) yet. Neither can r-hub. Are there
some Windows folks who can give it a spin (or who have given it a spin)?

Alternatively, an initial CRAN submission w/Linux & macOS as the two (out
of three) major platforms would also work, especially if the current
`harbor` API seems to be usable/logical.

TLDR: I wld not mind putting the finishing touches on it for a CRAN sub if
it's working for folks or just needs minor tweaks for folks.

On Tue, Sep 5, 2017 at 12:23 PM, Scott Chamberlain <notifications@github.com

**Comment by nuest** [comment]
:+1: on initial submission for Linux/macOS, though if Windows tests are a blocker, I'd get my hands on a Windows 10 machine and run tests.

**Comment by nuest** [comment]
The latest Appveyor has Docker support and I was able to let `harbor` tests run:

https://ci.appveyor.com/project/nuest/harbor/build/1.0.3

![image](https://user-images.githubusercontent.com/1325054/34603089-bf4145a4-f202-11e7-818e-517b3ef09c95.png)

The tests all run, see "artifacts" > file `harbor.Rcheck\tests\testthat.Rout`:

```
> library(testthat)
> library(harbor)
> 
> test_check("harbor")
== testthat results  ===========================================================
OK: 3 SKIPPED: 0 FAILED: 0
> 
```

I was also able to fix warnings and tests complete for different R versions:

https://ci.appveyor.com/project/nuest/harbor/build/1.0.5

![image](https://user-images.githubusercontent.com/1325054/34603843-dffc808a-f205-11e7-8f8c-ac560190b1d0.png)

------

So basic features work on Windows and can be tested on Appveyor with these changes: https://github.com/wch/harbor/compare/master...nuest:appveyor?expand=1

I'd be happy to contribute some more tests following the [todo list](https://github.com/hrbrmstr/harbor/blob/master/README.md) by @hrbrmstr and create a PR.

- [ ] Before CRAN, the status badges should be updated to the main repo.

**Comment by nuest** [comment]
@hrbrmstr bump.
I'm happy to contribute what's needed.

**Comment by hrbrmstr** [comment]
Oh this is gd news (apologies for the late reply). Have folks seen https://github.com/richfitz/stevedore ? It does the docker API vs cmdline machinations but is pure R vs an R shim to Python.

**Comment by nuest** [comment]
I did not know `stevedore`, thanks for the pointer. The approach using the API+Swagger is quite cool. AFAICS `stevedore` is pure R, but the required package [`httppipe` needs reticulate](https://github.com/richfitz/httppipe), at least currently, for Windows support.

@hrbrmstr _Do you suggest evaluating the different approaches of existing packages before going ahead with submitting `harbor` to CRAN?_
